### PR TITLE
ARROW-16782: [Format] Add REE definitions to FlatBuffers

### PR DESCRIPTION
**Pull request by zagto**

### CODE REVIEW

**Review by alamb — COMMENTED** [review]
FBS change looks good to me -- thank you @zagto 

Is it possible to incorporate the documentation from https://github.com/apache/arrow/pull/13333 into this pR as well so that the new changes are documented?

For anyone following along, see https://lists.apache.org/thread/539scy67qom5t2fkkd1m6fvh5htvwo3s discussion for more context

cc @zeroshade

**Comment by pitrou** [comment]
> Is it possible to incorporate the documentation from #13333 into this pR as well so that the new changes are documented?

+1 on this. It will also make reviewing easier.

**Comment by zeroshade** [comment]
@alamb @pitrou I rebased the docs from #13333 into this PR (to maintain the commit history from @zagto) and then updated them with the new name being "Run-End encoded" rather than "Run-Length encoded". Let me know if there's anything else needed to add.

File: docs/source/format/Columnar.rst
Comment (ianmcook) [inline docs/source/format/Columnar.rst:775]:
```suggestion
Run-end encoding (REE) is a variation of run-length encoding (RLE). Run-end
encoding is well-suited for representing data containing sequences of the same
value, called runs. In run-end encoding, each run is represented as a value
and an integer giving the index in the array where the run ends.
```

File: docs/source/format/Columnar.rst
Comment (ianmcook) [inline docs/source/format/Columnar.rst:788]:
```suggestion
index using binary search. The length of an individual run can be determined by
subtracting two adjacent values. (Contrast this with run-length encoding, in
which the lengths of the runs are represented directly, and in which random
access is less efficient.)
```

**Comment by zeroshade** [comment]
@github-actions crossbow submit preview-docs

File: format/Schema.fbs
Comment (lidavidm) [inline format/Schema.fbs:182]: Hmm, shouldn't this encode the run-end index size, like Int above or Decimal?

File: docs/source/format/Columnar.rst
Comment (lidavidm) [inline docs/source/format/Columnar.rst:782]:
```suggestion
64-bit integers. The actual values of each run are held in
the second child array.
```

File: format/Schema.fbs
Comment (westonpace) [reply format/Schema.fbs:182]: I think, because the ends are a child array, and not a buffer, that this is not strictly needed.  RunEndEncoded ens up being more similar to Struct array in this regard.

Comment (zeroshade) [reply format/Schema.fbs:182]: As it is currently implemented, the run-end index size isn't explicitly part of the type and is instead just an implementation detail in that the first child (containing the run-ends) can be any of int16/int32/int64. For example, note how List doesn't encode the child type and struct doesn't encode the child fields.

At least that's my understanding of why.

**Comment by zeroshade** [comment]
@github-actions crossbow submit preview-docs

**Comment by westonpace** [comment]
Do we want to prescribe a standard `name` for the ends and value children?  For example, with `List` some implementations use `item`.  Other implementations are free to use other names and, at times, this has in the past led to subtle incompatibilities (e.g. https://issues.apache.org/jira/browse/ARROW-14999)

**Comment by zeroshade** [comment]
@westonpace that's likely a good idea. The Go and C++ initial implementations that me and @zagto wrote settled on collectively calling them `run_ends` and `values`, but it would be good to explicitly prescribe them in the documentation. If no one objects to those names / the idea of prescribing them, I can update the doc to do so. I'll wait till tomorrow to give ppl a chance to weigh in before I update the `Columnar.rst` file

**Comment by lidavidm** [comment]
Yes, that's a good idea. Nice catch Weston.

**Review by alamb — COMMENTED** [review]
I reviewed the changes to docs/source/format/Columnar.rst and format/Schema.fbs and they looked good to me. Thank you for pushing this through

cc @tustvold

File: docs/source/format/Columnar.rst
Comment (ianmcook) [reply docs/source/format/Columnar.rst:775]: It looks like this and some other changes were reverted in the rebase

Comment (zeroshade) [reply docs/source/format/Columnar.rst:775]: yea, i noticed. I'm fixing it.

**Review by lidavidm — APPROVED** [review]

File: docs/source/format/Columnar.rst
Comment (pitrou) [inline docs/source/format/Columnar.rst:771]: You need to ensure that the decoration is exactly aligned to the title:
```suggestion
Run-End Encoded Layout
----------------------
```

File: docs/source/format/Columnar.rst
Comment (pitrou) [inline docs/source/format/Columnar.rst:781]: There seems to be some redundancy here
```suggestion
by itself, but has two child arrays. The first child array, called the run ends array,
holds either 16, 32, or 64-bit signed integers. The actual values of each run
are held in the second child array.
```

File: docs/source/format/Columnar.rst
Comment (pitrou) [inline docs/source/format/Columnar.rst:788]: This is confusing. Why start by saying they represent the length, and then explain that they don't hold the actual length?

File: docs/source/format/Columnar.rst
Comment (pitrou) [inline docs/source/format/Columnar.rst:795]:
```suggestion
run ends array all are positive and in strictly ascending order. A run end cannot be
```

File: docs/source/format/Columnar.rst
Comment (pitrou) [inline docs/source/format/Columnar.rst:808]: Is it "children arrays" or "child arrays"?

File: docs/source/format/Columnar.rst
Comment (pitrou) [inline docs/source/format/Columnar.rst:819]: This should indented the same as "run_ends" above.

File: docs/source/format/Columnar.rst
Comment (pitrou) [inline docs/source/format/Columnar.rst:815]:
```suggestion
          | Bytes 0-3   | Bytes 4-7   | Bytes 8-11  | Bytes 12-63           |
```

File: docs/source/format/Columnar.rst
Comment (pitrou) [inline docs/source/format/Columnar.rst:829]:
```suggestion
          | Bytes 0-3   | Bytes 4-7   | Bytes 8-11  | Bytes 12-63           |
```

File: format/Schema.fbs
Comment (pitrou) [inline format/Schema.fbs:181]: Can you add a short comment above to briefly explain this?

File: docs/source/format/Columnar.rst
Comment (jorisvandenbossche) [inline docs/source/format/Columnar.rst:812]: The run ends can have nulls?

File: docs/source/format/Columnar.rst
Comment (zeroshade) [reply docs/source/format/Columnar.rst:770]: Interestingly dictionary encoding isn't listed in that list of physical layouts, and REE is probably closest to that in terms of definitions. Should we also add dictionary encoded layout to this list?

File: docs/source/format/Columnar.rst
Comment (zeroshade) [reply docs/source/format/Columnar.rst:812]: there shouldn't be any nulls in the run_ends array, i'll make that explicit. Good catch

Comment (zeroshade) [reply docs/source/format/Columnar.rst:812]: Actually it's already stated just above these examples that a run end cannot be null, i'll just update this diagram to be explicit.

File: docs/source/format/Columnar.rst
Comment (zeroshade) [reply docs/source/format/Columnar.rst]: Is the updated version better?

File: docs/source/format/Columnar.rst
Comment (zeroshade) [reply docs/source/format/Columnar.rst]: updated to "Child Arrays"

File: docs/source/format/Columnar.rst
Comment (zeroshade) [reply docs/source/format/Columnar.rst]: fixed

File: format/Schema.fbs
Comment (zeroshade) [reply format/Schema.fbs:186]: Added

File: docs/source/format/Columnar.rst
Comment (pitrou) [reply docs/source/format/Columnar.rst]: Yes, thanks!

File: docs/source/format/Columnar.rst
Comment (jorisvandenbossche) [reply docs/source/format/Columnar.rst:770]: Dictionary encoding is a bit different in the specification since it's treated separately and not as a "Type", that's maybe the reason it's not listed there.

File: docs/source/format/Columnar.rst
Comment (jorisvandenbossche) [reply docs/source/format/Columnar.rst]: That might have been discussed before(and probably also part of the vote), but if they can't have nulls and thus basically are a single buffer, is there a reason that run_ends are not a "buffer" of the array instead of the first "child array" ? (like variable size list has offsets as buffer and values as child array)

File: docs/source/format/Columnar.rst
Comment (zeroshade) [reply docs/source/format/Columnar.rst:776]: I added both to that list, let me know what you think of the addition and if you think I should tweak it at all.

File: docs/source/format/Columnar.rst
Comment (jorisvandenbossche) [reply docs/source/format/Columnar.rst]: (have them as a buffer would avoid the ambiguity around the validity bitmap of the run_ends as child array)

Comment (zeroshade) [reply docs/source/format/Columnar.rst]: found the mailing list discussion on this which explicitly addresses being children arrays rather than a buffer: https://lists.apache.org/thread/bfz3m5nyf7flq7n6q9b1bx3jhcn4wq29 

TL;DR: in addition to the point I made above, there's this comment from @westonpace at the end of the thread:

> * There are two child arrays not necessarily for slicing but more to help
distinguish between the logical length (there are no buffers with the
logical length but there is an array) and the physical length (all of the
buffers are in these child arrays which are sized to the physical length).

Comment (lidavidm) [reply docs/source/format/Columnar.rst]: It might be good to preserve that reasoning somewhere more easily findable for posterity?

Comment (zeroshade) [reply docs/source/format/Columnar.rst]: added to `Columnar.rst`. Let me know if that's sufficient or if anyone thinks I should modify it a bit.

File: docs/source/format/Columnar.rst
Comment (westonpace) [inline docs/source/format/Columnar.rst:814]: I don't know if we want to justify our decision quite so much on the spec page.  Maybe something simpler like:

```suggestion
   Because the ``run_ends`` child array cannot have nulls, it's reasonable
   to consider why the ``run_ends`` are a child array instead of just a
   buffer, like the offsets for a :ref:`variable-size-list-layout`. This
   layout was considered, but it was decided to use the child arrays. 

   Child arrays allow us to keep the "logical length" (the decoded length)
   associated with the parent array and the "physical length" (the number
   of run ends) associated with the child arrays.  If ``run_ends`` was a
   buffer in the parent array then the size of the buffer would be unrelated
   to the length of the array and this would be confusing.
```

Comment (pitrou) [reply docs/source/format/Columnar.rst:814]: +1, perhaps we need some kind of separate FAQ

Comment (zeroshade) [reply docs/source/format/Columnar.rst:814]: you think it's not worth keeping the link to the original discussion in the spec?

Comment (pitrou) [reply docs/source/format/Columnar.rst:814]: We generally don't keep links to design discussions in the specs themselves, it would be weird to make a special case here.

Comment (zeroshade) [reply docs/source/format/Columnar.rst:814]: sounds good to me. Committing this suggestion.

**Review by westonpace — APPROVED** [review]

**Comment by lidavidm** [comment]
General question here: how are we going to handle versioning of the Arrow specification? Are we going to merge this and down the line release an "Arrow specification 1.1.0"/MetadataVersion.V6 once we also have types like StringView, or are we going to bump these versions for each new type?

**Comment by pitrou** [comment]
I think we only increase `MetadataVersion` when we make incompatible changes to the Flatbuffers definition, which is not the case here.

As for the format version, we seem to be inconsistent about it (no mention of buffer compression, for example):
```
/// Format Version History.
/// Version 1.0 - Forward and backwards compatibility guaranteed.
/// Version 1.1 - Add Decimal256 (No format release).
/// Version 1.2 (Pending)- Add Interval MONTH_DAY_NANO
```

**Comment by pitrou** [comment]
I don't think a new vote is needed, though I don't know for sure.

**Comment by lidavidm** [comment]
I suppose the version bump is implicit in adding a new data type? Looking at the Git history/the MONTH_DAY_NANO PR, the versions were just bumped when we added the type. (...And 1.2 is still "pending", guess we should take that out now and put 1.3 as adding REE.)

**Comment by zeroshade** [comment]
I've updated the `Schema.fbs` and `Columnar.rst` files for the version and regenerated files to pick up the new comments.

I've got two approvals here, so unless someone comments otherwise I'll merge this after EOD today, to give people a chance to add any final comments.

File: format/Schema.fbs
Comment (pitrou) [inline format/Schema.fbs:23]: I don't know if we want to keep those "No format release" comments. They were probably related to the idea of doing format releases separate from implementation releases, but it doesn't seem like we're planning to ever do that?

@lidavidm Thoughts?

Comment (lidavidm) [reply format/Schema.fbs:23]: I agree the comments are probably redundant. Implementations should just specify which format version they support.

Comment (zeroshade) [reply format/Schema.fbs:23]: okay, i'll remove those `(No format release)` comments.

**Review by alamb — APPROVED** [review]
I reviewed only the changes to `docs/source/format/Columnar.rst` and `format/Schema.fbs` but they look great to me. Thank you to everyone who has been working on this proposal -- it is super exciting to see it so close

**Pull request merged**
